### DISCUSSION TITLE
How to use truth_space_table_from_labels_column()? Contradictory API.

### DISCUSSION THREAD

**Discussion by aalexandersson** [moj-analytical-services/splink#2045] [Q&A] [answered]
The API for [Linker.truth_space_table_from_labels_column](https://moj-analytical-services.github.io/splink/linkermodelviz.html#splink.linker.Linker.truth_space_table_from_labels_column) requires parameter `labels_column_name` in the header and description section but it instead contradictory requires the parameter `labels_tablename` in its parameter section. Is the required parameter `labels_column_name` or `labels_tablename`? 

The API-provided example is `linker.truth_space_table_from_labels_column("cluster")`
What is the required data format to reproduce the example? I tried this code on my test data:

```
labels = pd.read_csv("tmp_labels.csv")
linker.register_table(labels, "labels", overwrite=True)
linker.query_sql("select * from labels")
```

The last command gives this output:
```
<splink.duckdb.linker.DuckDBDataFrame object at 0x000001A75B242ED0>
         simulant_id  cluster
0       1007_1001150        1
1       1007_1001333        1
2       1007_1006629        1
3       1007_1006679        1
4       1007_1009212        1
...              ...      ...
670209     99_810000        0
670210     99_828733        0
670211     99_902773        0
670212     99_925429        0
670213     99_939553        0

[670214 rows x 2 columns]
```
Is that the required data format?

Here, for the above partial test data, I tried this command:
`linker.truth_space_table_from_labels_column("labels")`

But it gives the error:
`Binder Error: Values list "l" does not have a column named "labels"`

Similarly, I now tried this command:
`linker.truth_space_table_from_labels_column("cluster")`

But it gives the error:
`Binder Error: Values list "l" does not have a column named "cluster"`

Background context: I have ground truth data, as above where cluster=1 means an actual match, and cluster=0 means an actual non-match. [PR 1478 (Robin)](https://github.com/moj-analytical-services/splink/pull/1478#issuecomment-1659737882) suggested I use `truth_space_table_from_labels_column()`. But how do I do that since the API is contradictory and my code resulted in error? I used Splink 3.9.13 on Windows 10. I need to use ground truth data to calculate performance measures such as the confusion matrix.

Update 1: Using the standard variable name `unique_id` instead of `simulant_id` for the unique id values leads to the same error.
Update 2: I guess that the variable `cluster` should have more values. There, for troubleshooting I am now going back to basics and instead using the known table format. I tried this code:
```
import duckdb
labels = duckdb.sql(
    """
select * from read_csv_auto("tmp_labelstable.csv")
"""
).df()
labels = linker.register_labels_table(labels)
linker.roc_chart_from_labels_table(labels)
```
The last line `linker.roc_chart_from_labels_table(labels)` gives an error "list index out of range". Here is the end of the trace back:
```
splink\charts.py", line 219, in roc_chart
    r = records[0]
        ~~~~~~~^^^
IndexError: list index out of range
```
I get the same error message but for a different line in `charts.py` if I instead try 
`linker.precision_recall_chart_from_labels_table(labels)`. The error message is:
```
line 237, in precision_recall_chart
    r = records[0]
        ~~~~~~~^^^
IndexError: list index out of range
```

I get no error message if I instead try:
`linker.truth_space_table_from_labels_table(labels)` . The output is not usable yet though:
`<splink.duckdb.linker.DuckDBDataFrame object at 0x000001A68FA6AA10>`
In summary, I seem to make some progress but the user documentation is not sufficient.

Update 3: Dead link [here](https://moj-analytical-services.github.io/splink/linkerqa.html#splink.linker.Linker.roc_chart_from_labels) for the required format of the labels file. See [tutorials/07_Evaluation](https://moj-analytical-services.github.io/splink/demos/tutorials/07_Evaluation.html#load-in-labels).

Update 4: I can access the labels_table but the truth_space_table is empty. I used this code:
```
import duckdb
labels = duckdb.sql(
    """
select * from read_csv_auto("tmp_labelstable2.csv")
"""
).df()

labels.head(5)
labels_table = linker.register_labels_table(labels)

linker.truth_space_table_from_labels_table(labels_table) 
labels_table.as_record_dict(limit=2)  # -> 2 records

tt = linker.truth_space_table_from_labels_table(labels_table) 
tt.as_pandas_dataframe(limit=5)  # -> empty dataframe, why?
```

This is the console output:
```
>>> import duckdb
>>> labels = duckdb.sql(
...     """
... select * from read_csv_auto("tmp_labelstable2.csv")
... """
... ).df()
>>> 
>>> labels.head(5)
  simulant_id_l source_dataset_l  ... source_dataset_r clerical_match_score
0  1007_1001150              df1  ...              df2                    1
1  1007_1001333              df1  ...              df2                    1
2  1007_1006629              df1  ...              df2                    1
3  1007_1006679              df1  ...              df2                    1
4  1007_1009212              df1  ...              df2                    1

[5 rows x 5 columns]
>>> labels_table = linker.register_labels_table(labels)
>>> 
>>> linker.truth_space_table_from_labels_table(labels_table) 
<splink.duckdb.linker.DuckDBDataFrame object at 0x000002A6187906D0>
>>> labels_table.as_record_dict(limit=2)  # -> 2 records
[{'simulant_id_l': '1007_1001150', 'source_dataset_l': 'df1', 'simulant_id_r': '1007_1001150', 'source_dataset_r': 'df2', 'clerical_match_score': 1}, {'simulant_id_l': '1007_1001333', 'source_dataset_l': 'df1', 'simulant_id_r': '1007_1001333', 'source_dataset_r': 'df2', 'clerical_match_score': 1}]
>>> 
>>> tt = linker.truth_space_table_from_labels_table(labels_table) 
>>> tt.as_pandas_dataframe(limit=5)  # -> empty dataframe
Empty DataFrame
Columns: [truth_threshold, match_probability, row_count, p, n, tp, tn, fp, fn, P_rate, N_rate, tp_rate, tn_rate, fp_rate, fn_rate, precision, recall, specificity, npv, accuracy, f1, f2, f0_5, p4, phi]
Index: []
>>> 
```

If I use the format as described at [truth_space_table_from_labels_table](https://moj-analytical-services.github.io/splink/linker.html?h=truth_space_table_from_labels_table#splink.linker.Linker.truth_space_table_from_labels_table), that is, with the first variable being `source_dataset_l`, then the result is the same empty DataFrame.

Update 5: The issue is almost solved. As documented at [truth_space_table_from_labels_table](https://moj-analytical-services.github.io/splink/linker.html?h=truth_space_table_from_labels_table#splink.linker.Linker.truth_space_table_from_labels_table), emphasis added:
Note that `source_dataset` and `unique_id` should correspond to the values specified in the settings dict, and the **input_table_aliases** passed to the linker object.

The only remaining issue is that the output shows 0 false positives (FP) and 0 true negatives (TN), but my test data has a few FP and almost 10,000 TN.

**Comment by RobinL** [comment, marked as answer]
Hi @aalexandersson .  Sorry for late response.  Yeah, that's right - it looks like there's a typo in the docstring [here](https://github.com/moj-analytical-services/splink/blob/65f3a78b8b77dddbb226eee808f304f068bd22a3/splink/linker.py#L2886) where labels_tablename should read labels_column_name

<details>
<summary>
Reprex of it working, click to expand
</summary>

```python
from splink.datasets import splink_dataset_labels, splink_datasets
from splink.duckdb.blocking_rule_library import block_on
from splink.duckdb.comparison_library import (
    exact_match,
    levenshtein_at_thresholds,
)
from splink.duckdb.linker import DuckDBLinker

df = splink_datasets.fake_1000

linker = DuckDBLinker(df)

df = splink_datasets.fake_1000

settings = {
    "probability_two_random_records_match": 0.01,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        block_on(["first_name"]),
        block_on(["surname"]),
    ],
    "comparisons": [
        levenshtein_at_thresholds("first_name", 2),
        exact_match("surname"),
        exact_match("dob"),
        exact_match("city", term_frequency_adjustments=True),
        exact_match("email"),
    ],
    "retain_intermediate_calculation_columns": True,
    "additional_columns_to_retain": ["cluster"],
}


linker = DuckDBLinker(df, settings)
# train model here
df_predictions = linker.predict(threshold_match_probability=0.2)

linker.truth_space_table_from_labels_column("cluster").as_pandas_dataframe()

```

</details>

I think the second issue is a bug here, which I realised we'd found before but failed to raise.  I've now created it here:
https://github.com/moj-analytical-services/splink/issues/2059

The main reason this seems to have slipped though the cracks is that it's rarely used, because people almost never have fully labelled datasets.

For small datasets, i _think_ you might get accurate stats if you block on `1=1` i.e. cartesian product, but obviously that's not feasilble in most applications because it's computationally intractible
